compiler/rustc_hir_typeck/src/expr.rs

```diff
@@ -247,16 +247,39 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             // diverging expression (e.g. it arose from desugaring of `try { return }`),
             // we skip issuing a warning because it is autogenerated code.
             ExprKind::Call(..) if expr.span.is_desugaring(DesugaringKind::TryBlock) => {}
-            ExprKind::Call(callee, _) => self.warn_if_unreachable(expr.hir_id, callee.span, "call"),
+            ExprKind::Call(callee, _) => {
+                let emit_warning = if let ExprKind::Path(ref qpath) = callee.kind {
+                    // Do not emit a warning for a call to a constructor.
+                    let res = self.typeck_results.borrow().qpath_res(qpath, callee.hir_id);
+                    !matches!(res, Res::Def(DefKind::Ctor(..), _))
```

Comment:
i am a bit confused why that doesn't warn :thinking: 

> Constructors are carved out on purpose, to avoid an "unreachable call" diagnostic on them, which is not very useful as they are inert.

I don't understand that comment :thinking: I would assume the outer constructor to keep warning here?
```rust
struct Foo<T>(T);

fn foo() -> Foo<String> {
    Foo(return Foo(String::new()))
}
```